tools/generate_release_notes.py

```diff
@@ -1,246 +1,331 @@
-"""Generate the release notes automatically from Github pull requests.
-
-Start with:
-```
-export GH_TOKEN=<your-gh-api-token>
-```
-
-Then, for a major release:
-```
-python /path/to/generate_release_notes.py v0.14.0 main --version 0.15.0
-```
-
-For a minor release:
-```
-python /path/to/generate_release_notes.py v.14.2 v0.14.x --version 0.14.3
-```
-
-You should probably redirect the output with:
-```
-python /path/to/generate_release_notes.py [args] | tee release_notes.rst
-```
-
-You'll require PyGitHub and tqdm, which you can install with:
-```
-pip install -r requirements/_release_tools.txt
-```
-
-References
-https://github.com/scikit-image/scikit-image/issues/3404
-https://github.com/scikit-image/scikit-image/issues/3405
-"""
+"""Generate release notes automatically from GitHub pull requests."""
+
+
 import os
+import re
+import sys
 import argparse
-from datetime import datetime
-from warnings import warn
-
-from github import Github
-
-try:
-    from tqdm import tqdm
-except ImportError:
-    warn(
-        "tqdm not installed. This script takes approximately 5 minutes "
-        "to run. To view live progressbars, please install tqdm. "
-        "Otherwise, be patient."
-    )
+import logging
+from pathlib import Path
+from dataclasses import dataclass
+from typing import Callable
+from collections.abc import Iterable
+
+import requests_cache
+from tqdm import tqdm
+from github import Github, Repository, PullRequest, NamedUser, Commit
 
-    def tqdm(i, **kwargs):
-        return i
 
+logger = logging.getLogger(__name__)
 
-GH_USER = "scikit-image"
+here = Path(__file__).parent
+
+GH_URL = "https://github.com"
+GH_ORG = "scikit-image"
 GH_REPO = "scikit-image"
-GH_TOKEN = os.environ.get("GH_TOKEN")
-if GH_TOKEN is None:
-    raise RuntimeError(
-        "It is necessary that the environment variable `GH_TOKEN` "
-        "be set to avoid running into problems with rate limiting. "
-        "One can be acquired at https://github.com/settings/tokens.\n\n"
-        "You do not need to select any permission boxes while generating "
-        "the token."
-    )
 
-g = Github(GH_TOKEN)
-repository = g.get_repo(f"{GH_USER}/{GH_REPO}")
 
+def lazy_tqdm(*args, **kwargs):
+    """Defer initialization of progress bar until first item is requested."""
```

Comment:
Does the progress bar take long to initialize?

Reply:
Not at all, the problem is that it initalizes right after the call to `tqdm()` at doesn't wait until the iteration is actually started. Which messes up the formatting here https://github.com/scikit-image/scikit-image/blob/440f2d4c9be3384f0d4856261feebab67b07d736/tools/generate_release_notes.py#L304-L310

I do it this way because I like to handle script output in script specific functions and keep it out of re-usable ones like `contributors()` (logging being the configurable exception). That's why I am not calling `tqdm` in `contributors()` instead but wrapping the input. I'll try to expand the docstring to justify it's existence.